Source/FixedBackground/KerrSchildFixedBG.hpp

```diff
@@ -0,0 +1,275 @@
+    /// Work out the gradients of the quantities H and el appearing in the Kerr
+    /// Schild solution
+    template <class data_t>
+    void get_KS_derivs(Tensor<1, data_t> &dHdx, Tensor<2, data_t> &dldx,
+                       Tensor<1, data_t> &dltdx, const data_t &H,
+                       const Coordinates<data_t> &coords) const
+    {
+        // black hole params - mass M and boost v
+        const double M = m_params.mass;
+        const double a = m_params.spin;
+        const double a2 = a * a;
+
+        // work out where we are on the grid, and useful quantities
+        Tensor<1, data_t> x;
+        x[0] = coords.x;
+        x[1] = coords.y;
+        x[2] = coords.z;
+        const double z = coords.z;
+        const data_t rho = coords.get_radius();
+        const data_t rho2 = rho * rho;
+
+        // the Kerr Schild radius r
+        const data_t r2 = 0.5 * (rho2 - a2) +
+                          sqrt(0.25 * (rho2 - a2) * (rho2 - a2) + a2 * z * z);
+        const data_t r = sqrt(r2);
+        const data_t cos_theta = z / r;
+        const data_t cos_theta2 = cos_theta * cos_theta;
+
+        using namespace TensorAlgebra;
+        // derivatives of r wrt actual grid coords
+        Tensor<1, data_t> drhodx;
+        FOR1(i) { drhodx[i] = x[i] / rho; }
+
+        Tensor<1, data_t> drdx;
+        FOR1(i)
+        {
+            drdx[i] =
+                0.5 / r *
+                (rho * drhodx[i] +
+                 0.5 / sqrt(0.25 * (rho2 - a2) * (rho2 - a2) + a2 * z * z) *
+                     (drhodx[i] * rho * (rho2 - a2) +
+                      delta(i, 2) * 2.0 * a2 * z));
+        }
+
+        Tensor<1, data_t> dcosthetadx;
+        FOR1(i) { dcosthetadx[i] = -z / r2 * drdx[i] + delta(i, 2) / r; }
+
+        FOR1(i)
+        {
+            dHdx[i] = H * (drdx[i] / r -
+                           2.0 / (r2 + a2 * cos_theta2) *
+                               (r * drdx[i] + a2 * cos_theta * dcosthetadx[i]));
+        }
+
+        // note to use convention as in rest of tensors the last index is the
+        // derivative index so these are d_i l_j
+        FOR1(i)
+        {
+            // first the el_x comp
+            dldx[0][i] =
+                (x[0] * drdx[i] + r * delta(i, 0) + a * delta(i, 1) -
+                 2.0 * r * drdx[i] * (r * x[0] + a * x[1]) / (r2 + a2)) /
+                (r2 + a2);
+            // now the el_y comp
+            dldx[1][i] =
+                (x[1] * drdx[i] + r * delta(i, 1) - a * delta(i, 0) -
+                 2.0 * r * drdx[i] * (r * x[1] - a * x[0]) / (r2 + a2)) /
+                (r2 + a2);
+            // now the el_z comp
+            dldx[2][i] = -x[2] * drdx[i] / r2 + delta(i, 2) / r;
+        }
+
+        // then dltdi
+        FOR1(i) { dltdx[i] = 0.0; }
+    }
+
+  public:
+    // used to decide when to excise - ie when within the horizon of the BH
+    // note that this is not templated over data_t
+    double excise(const Cell<double> &current_cell) const
```

Comment:
Maybe here also pass coordinates? And change the return object - here we have two horizons so this becomes ambiguous

Reply:
Here I changed this again to bool, that excises if the outer horizon is <1. Is that right?